[Core] Fix torch.dynamo compatibility for Qwen models on vllm-gaudi

pawel-olejniczak opened the pull request:
<!-- markdownlint-disable -->

## Purpose

Fix critical `torch.dynamo` compatibility issues that prevent Qwen2.5 models from running on vllm-gaudi.

This PR addresses three interconnected issues identified during investigation:
1. **Dispatch key set mismatch**: `make_tensor_with_pad` used `torch.from_numpy()`, creating tensors incompatible with `torch.dynamo` compilation
2. **Builtin function compilation failure**: `torch.dynamo` couldn't compile `max(..., default=0)` syntax  
3. **AttributeError on None inputs**: Penalty application logic wasn't robust to `None` inputs during warmup scenarios

**Error Messages Fixed:**
- `AssertionError: Guard check failed! tensor_from_numpy(stack[0]): dispatch key set mismatch`
- `incorrect arg count <bound method BuiltinVariable._call_min_max of BuiltinVariable(max)> got an unexpected keyword argument 'default'`
- `AttributeError: 'NoneType' object has no attribute 'device'`

## Test Plan

### Local Testing Environment
- **Hardware**: Intel Gaudi3 accelerator via vllm-gaudi
- **Model**: `Qwen/Qwen2.5-14B-Instruct`
- **Command**: 
`vllm serve Qwen/Qwen2.5-14B-Instruct --max-model-len 4096 --max-num-seqs 32`
- **Test Request**:
```
curl http://localhost:8000/v1/completions
-H "Content-Type: application/json"
-d '{
"model": "Qwen/Qwen2.5-14B-Instruct",
"prompt": "Give me a short introduction to large language model.",
"max_tokens": 100
}'
```

## Test Result

### Before Fix
```
AssertionError: Guard check failed!
160 tensor_from_numpy(stack): dispatch key set mismatch.
expected=DispatchKeySet(CPU, BackendSelect, ADInplaceOrView),
actual=DispatchKeySet(CPU, BackendSelect)
```

Server crashed on first inference request.

### After Fix
```
{
"id":"cmpl-421571a6c96049ad91448cfa679c34bc",
"object":"text_completion",
"created":1758727612,
"model":"Qwen/Qwen2.5-14B-Instruct",
"choices":[{
"index":0,
"text":" "A large language model (LLM) is an advanced...",
"logprobs":null,
"finish_reason":"length"
}],
"usage":{"prompt_tokens":16,"total_tokens":116,"completion_tokens":100}
}
```

### Changes Summary
1. **Refactored `make_tensor_with_pad`** (vllm/utils/__init__.py): Pure PyTorch implementation eliminates NumPy interop issues
2. **Enhanced penalty logic robustness** (vllm/model_executor/layers/utils.py): Added None-safety to prevent AttributeErrors
3. **Fixed compiler compatibility**: Replaced `max(..., default=0)` with conditional expression

---

**Essential Elements Checklist:**
- [x] **Purpose**: Fix torch.dynamo compatibility for Qwen models on vllm-gaudi
- [x] **Test plan**: Local testing with Qwen/Qwen2.5-14B-Instruct on vllm-gaudi
- [x] **Test results**: Before/after comparison showing successful inference
- [x] **Documentation**: Not required - internal compatibility fix
- [x] **Release notes**: Not required - bug fix improving existing functionality

pawel-olejniczak committed:
[Core] Fix torch.dynamo compatibility for Qwen models

- Replace NumPy-based make_tensor_with_pad with pure-PyTorch implementation
- Add None-safety to penalty application functions
- Fix max() builtin compilation issues with torch.dynamo
- Enables Qwen2.5-14B-Instruct to run successfully on vllm-gaudi

Fixes compatibility issues where torch.dynamo guard failures occurred
due to dispatch key set mismatches and AttributeErrors when applying
repetition penalties to logits.

Signed-off-by: Paweł Olejniczak <polejniczakx@habana.ai>

ProExpertProg reviewed (commented):
Looks reasonable to me but not familiar with this part of the code.

pawel-olejniczak committed:
Merge branch 'main' into fix/torch-dynamo-compatibility-qwen2

Signed-off-by: Paweł Olejniczak <polejniczakx@habana.ai>

pawel-olejniczak committed:
[Core] Refine torch.compile fixes for upstream changes
Adapt to upstream refactoring with pure-PyTorch tensor creation:
- Rewrite make_tensor_with_pad with torch.full instead of NumPy
- Add None-safety to apply_penalties
- Remove redundant defensive checks
- Fix max() builtin torch.dynamo compatibility

Signed-off-by: Paweł Olejniczak <polejniczakx@habana.ai>

pawel-olejniczak committed:
Merge branch 'main' into fix/torch-dynamo-compatibility-qwen2

pawel-olejniczak commented on vllm/entrypoints/openai/tool_parsers/minimax_m2_tool_parser.py:
```diff
@@ -2,11 +2,12 @@
 from collections.abc import Sequence
 from typing import Any
 
+import regex as re
```
It looks like someone has already fixed it.

pawel-olejniczak committed:
Merge branch 'main' into fix/torch-dynamo-compatibility-qwen2

pawel-olejniczak commented:
I fixed the issue on the vllm-gaudi side.
